### ISSUE TITLE
Threading and Avalonia properties

### ISSUE THREAD

**Issue by wgebczyk** [AvaloniaUI/Avalonia#2576]
First thanks for such big effort! :)

TLDR: The code of at least AvaloniaProperty seems to be not thread safe, with hidden assumption that only main UI thread might interact with it.

When looking at the code of AvaloniaProperty, I get impression that this code is expected to be called ONLY from one thread at same time. No checks are done for concurrency.
At least _metadata* are not thread safe structures. If there would be at least 2 Dispatchers involved and on them 2 property registration would happen, this might lead to undefined behavior (works, infinite loops, NRE, mess up state).

When looking at Dispatcher code and comment "In Avalonia, there is usually only a single 'Dispatcher' in the application" with public ctor it might indicate more than one Dispatcher.
On other hand, CheckAccess/VerifyAccess are using one single well known UI thread.

That's perfectly fine that AvaloniaUI will be single threaded UI framework. The HTML/JS/DOM proven that single threaded UI just works.
Still I would like to made it clear and appropriate code is put in different places to avoid accidental cross thread calls for the following reasons:
- as behavior is undefined for cross thread calls, the diagnostics of such issues is very hard
- this would cut all the bizarre ideas like off loading some UI work to other threads as they are clearly ot supported
- with await/async + Task it is very easy to land on not appropriate thread, hence ANY diagnostics are welcome.
- if such checks are not necessary (it is obvious it is single threaded), then why there is dispatcher concept with CheckAccess?

Of course it AvaloniaUI would be multi-dispatcher, then probably rather big refactor is needed.

So, what's the team stance on this topic?

**Comment by kekekeks** [comment]
Dispatchers were initially designed against WPF's dispatchers, but it turned out that most platforms don't allow multithreaded access to UI-related APIs. So the framework now enforces everything to work on the UI thread. We could implement support for multiple dispatchers, but that will only work for Win32 and X11.

>with public ctor it might indicate more than one Dispatcher.

I think at some point someone tried to reuse the displatcher code to run their own (non-UI) main loop on non-UI thread. It was quite a while ago though. We should probably remove it.

**Comment by Gillibald** [comment]
Hmm. Setting AvaloniaProperty always involves a call to VerifyAccess that throws if someone tries to get or set the property from a thread other than the ui thread.

You can always perform non ui related tasks on a different thread you just need to use the ui thread's dispatcher to marshall the result to the ui thread. That works the same with Wpf.

**Comment by wgebczyk** [comment]
@kekekeks That's fair.

@Gillibald Does the registration of property also does this check? Non concurrent versions of structures (Dictionary<>, Queue<>) without sync primitives looks suspicious.
But from your answer it seems that this is an assumption that AvaloniaUI is single threaded…
… and you have it under full control. That's good!

Thanks!
